packages/compose/src/hooks/use-modified-style/index.js

```diff
@@ -0,0 +1,122 @@
+ *
+ * @example
+ * ```
+ * selectModifiedStyles(
+ *   {
+ *    'block__element': { width: '10px' },
+ *    'block__element--narrow': { width: '1px' },
+ *    'block__element--wide': {width: '100px'}
+ *   },
+ *   'wide'
+ * );
+ * // returns { block_element': { width: '100px' } }
+ * ```
+ *
+ * @param {Object} styles
+ * @param {string} modifier - the base modifier without the `--` prefix
+ *
+ * @return {Object} - Subset of styles based on the modifier
+ */
+function selectModifiedStyles( styles, modifier ) {
+	const modifierSelectors = Object.keys( styles ).filter( ( selector ) =>
+		selector.match( `--${ modifier }$` )
+	);
+
+	return modifierSelectors.reduce( ( modifiedStyles, modifierSelector ) => {
+		const blockElementSelector = modifierSelector.split( '--' )[ 0 ];
+		modifiedStyles[ blockElementSelector ] = styles[ modifierSelector ];
+		return modifiedStyles;
+	}, {} );
+}
+
+/**
+ * Updates a styles object with modified styles if the update conditions are met.
+ *
+ * @example
+ *
+ * ```
+ * updateStyles(
+ *  {
+ *    'block__element-A' : { width: '10px' },
+ *    'block__element-B' : { 'font-family': 'Comic Sans' },
+ *    'block__element-C' : { visibility : 'hidden' }
+ *  },
+ *  {
+ *    'block__element-A' : { height: '10px' },
+ *    'block__element-B' : { 'font-family' : 'Helvetica' },
+ *    'block__element-Z' : { 'z-index' : 2147483647 }
+ *  },
+ *  [ true, true ]
+ * );
+ * // returns
+ * // {
+ * //   'block_element-A' : { width: '10px', height: '10px' },
+ * //   'block-element-B' : { 'font-family' : 'Helvetica' },
+ * //   'block-element-C' : { visibility : 'hidden' }
+ * // }
+ * ```
+ * @param {Object} styles
+ * @param {Object} styleUpdates
+ * @param {Array<boolean>} updateConditions - Boolean conditions on when to apply updates
+ *
+ * @return {Object} - Updated styles object
+ */
+function updateStyles( styles, styleUpdates, updateConditions ) {
+	// Test if the modifier conditions are met
+	const shouldUpdate = updateConditions.every( ( should ) => should );
+
+	if ( shouldUpdate ) {
+		Object.keys( styleUpdates ).forEach( ( selector ) => {
+			styles[ selector ] = {
+				...styles[ selector ],
+				...styleUpdates[ selector ],
+			};
+		} );
+	}
+	return styles;
+}
+
+function useModifiedStyle( baseStyles, modifiers ) {
+	// Memoize the modifier selectors
+	const modifierSelectors = useMemo( () => {
+		return Object.keys( modifiers );
+	}, [] );
+
+	// Memoize the modified styles
+	const modifiedStyles = useMemo( () => {
+		const _modifiedStyles = {};
+		for ( const modifier in modifiers ) {
+			_modifiedStyles[ modifier ] = selectModifiedStyles(
+				baseStyles,
+				modifier
+			);
+		}
+		return _modifiedStyles;
+	}, [] );
+
+	const futureStyles = { ...baseStyles };
+
+	// Apply the modifiers to the future styles
+	modifierSelectors.forEach( ( modifier ) => {
+		updateStyles(
+			futureStyles,
+			modifiedStyles[ modifier ],
+			modifiers[ modifier ]
```

Comment:
Our use of `useMemo` in this function seems a bit off to me since here we're using teh memorized keys of the `modifier` object (`modifiedSelectors`) but we're using those memorized keys on the most recently passed `modifiers` object. As long as those keys never change (and they probably won't) we'll be fine, but if they change we could end up passing undefined values (or missing new keys). 

I was playing around with this, and I did figure out [one option for how we could just memorize this entire function call](https://github.com/WordPress/gutenberg/commit/557ee25b389bda009bd1821cab6c3209915ba741). I also made a few other changes that I think might simplify the logic a bit, but don't feel like you have to take those changes, they're just suggestions/ideas.